### PR TITLE
feat: expire messages from the cache based on last seen time

### LINKED ISSUE #502: seenMessages TTL should be based on last observed time, not first
The [seenMessage](https://github.com/libp2p/go-libp2p-pubsub/blob/01ab84a809990a5ceefd34bca98139617fcc5389/pubsub.go#L152) cache prevents us from processing and re-broadcasting a pubsub message that we have already seen within the `seenMsgTTL` (default 2 minutes).  Accessing an existing message, however, [does not update the timestamp associated with the entry in the cache](https://github.com/whyrusleeping/timecache/blob/cfcb2f1abfee846c430233aef0b630a946e0a5a6/timecache.go#L57-L60).

This means that if a single pubsub message is being received frequently and consistently, we'll re-broadcast it every `seenMsgTTL`, rather than never re-broadcasting it as is the desired behavior.  If every time we received a duplicate pubsub message we pushed back the expiration time in the cache, that would give us the desired behavior of never re-broadcasting a message that we have already seen within the last `seenMsgTTL` seconds.

### PR DESCRIPTION
**Pull request by smrz2001**
This PR attempts to fix https://github.com/libp2p/go-libp2p-pubsub/issues/502.

It also includes a thread-safe copy of the original time cache [implementation](https://github.com/whyrusleeping/timecache).

Both the "first seen" (original) and "last seen" (new) implementations have been abstracted behind the `TimeCache` interface in a new `timecache` package.

cc @vyzo

### CODE REVIEW

File: blacklist.go
Comment (smrz2001) [inline blacklist.go:37]: Since the ported time cache implementation is thread-safe, I thought it would be ok to remove the `RWMutex` from here.

Comment (smrz2001) [reply blacklist.go:37]: There is a minor change in the atomicity of operations though. Before, the lock was held across the entire `TimeCachedBlacklist.Add()` call, and thus across both the `tc.Has()` and `tc.Add()` calls.

Now, the time cache lock will be held only across the individual `tc.Has()` and `tc.Add()` calls. I think that should be ok, but I can totally just revert this change.

File: go.mod
Comment (smrz2001) [inline go.mod:7]: I only added this package, the other changes are from `go mod tidy`.

File: pubsub.go
Comment (smrz2001) [inline pubsub.go]: As you suggested, I've defaulted the strategy to be "last seen", i.e. the new implementation.

File: blacklist.go
Comment (smrz2001) [reply blacklist.go:37]: One more thing - I retained the original (first-seen) time cache implementation here, but could also switch it to the new one.

**Comment by vyzo** [comment]
thanks, i was on vacation. I will take a look soon.

File: timecache/first_seen_cache.go
Comment (smrz2001) [inline timecache/first_seen_cache.go:70]: The expiration check was [absent](https://github.com/whyrusleeping/timecache/blob/master/timecache.go#L59) in the original implementation, which was a bug I [discovered](https://github.com/ipfs/kubo/pull/9372/files#diff-a75543978c37023721256291000855b02598e5d72668e203f7372cc4b8c159cfR222) while implementing the `SeenMessagesTTL` option in `kubo`.

I also added test cases for both cache implementations covering this scenario.

**Comment by BigLep** [comment]
@vyzo : after this is reviewed/merged, can you cut a new release?   (I see this repo isn't part of unified ci: https://github.com/protocol/.github/blob/master/configs/go.json).  I'm asking so it can be bubbled up to Kubo for https://github.com/ipfs/kubo/pull/9372/files#diff-a75543978c37023721256291000855b02598e5d72668e203f7372cc4b8c159cfR222

**Comment by vyzo** [comment]
sure

**Comment by BigLep** [comment]
@vyzo : I know you're juggling a lot.  Would it be helpful if someone from the EngRes libp2p Stewards team takes a look?

**Comment by vyzo** [comment]
It has been a little hectic in fvm land recently, but i do have some time to look at this during the weekend.

**Comment by vyzo** [comment]
also, feel free to add steward reviewers, it would help disseminate knowledge.

File: pubsub.go
Comment (lidel) [reply pubsub.go]: @smrz2001  bit confusing:  the old default is still present in 
https://github.com/libp2p/go-libp2p-pubsub/blob/04bfcf58514f59bddb650fec5b2edac018c8c406/timecache/time_cache.go#L19
and 
https://github.com/libp2p/go-libp2p-pubsub/blob/04bfcf58514f59bddb650fec5b2edac018c8c406/timecache/time_cache.go#L28-L30

**Review by lidel — DISMISSED** [review]
@vyzo I eyeballed it, but any chance you take a look at this week? 

@smrz2001 quick ask:  unify default strategy (right now it is different https://github.com/libp2p/go-libp2p-pubsub/pull/513/files#r1071344615)

**Comment by vyzo** [comment]
yes, yes, my weekend plan didnt materialize, but i have it on the list.

File: pubsub.go
Comment (smrz2001) [reply pubsub.go]: Good point! Fixed.

File: blacklist.go
Comment (smrz2001) [reply blacklist.go:37]: > One more thing - I retained the original (first-seen) time cache implementation here, but could also switch it to the new one.

Ok, scratch this. As per @lidel's comment, I aligned everything to use the new implementation.

**Comment by smrz2001** [comment]
> @vyzo I eyeballed it, but any chance you take a look at this week?
> 
> @smrz2001 quick ask: unify default strategy (right now it is different https://github.com/libp2p/go-libp2p-pubsub/pull/513/files#r1071344615)

Fixed! Thanks, @lidel.

**Comment by BigLep** [comment]
@MarcoPolo @marten-seemann @Wondertan : any chance you can please help take a look at this?  I'd like to get this over the line to get @smrz2001 and team unblocked.

**Comment by Wondertan** [comment]
Happy to review this on the weekends if it will be relevant until then.

File: pubsub.go
Comment (vyzo) [reply pubsub.go]: I would prefer to keep the existing behaviour as default.

I dont like surprising users with subtle changes, esp since this might effect memory usage.

Note that kubo is a fringe user, not surprising filecoin and eth is much more important here.

File: timecache/first_seen_cache.go
Comment (vyzo) [inline timecache/first_seen_cache.go:36]: This should probably happen in the backround instead of every add, it is inefficient.

We dont have to do it in this pr, but if we dont we should add a TODO and open an issue for continuation.

File: timecache/last_seen_cache.go
Comment (vyzo) [inline timecache/last_seen_cache.go:44]: same here.

File: timecache/last_seen_cache.go
Comment (vyzo) [inline timecache/last_seen_cache.go:73]: RLock?

**Review by vyzo — COMMENTED** [review]
A couple of issues to decide and a blocker: please dont change the default behaviour!

**Comment by vyzo** [comment]
Please retain the default behaviour of first seen.

**Comment by oed** [comment]
> Please retain the default behaviour of first seen.

@vyzo you said previously (https://github.com/libp2p/go-libp2p-pubsub/issues/502#issuecomment-1292586489) that it should be changed. This was the point of the bug that was reported.

**Comment by vyzo** [comment]
Yes, i know, but i had achange of heart regarding the _default_ behaviour.

It is true that it is desirable to use the last seen behaviour for many applications.

However, some of the biggest users of this library are blockchains which can quite easily post reject and deduplicate messages once the timecache period has expired, because of consensus.
That is, after two minutes almost all blockchains can reject older messages reappearing and replays with their consensus based validators (nonce, block height, and so on).

For these applications, changing the default under their feet will not bring any tangible benefit for the increased memory usage.
At the very least i think we should be conservative and let the user choose.

Note that we can also consider making last seen the default a bit further down the line, say in a release or two.
Also, maturity of implementation concerns.

Does this make sense?

**Comment by oed** [comment]
It does make sense. Thanks for explaining your reasoning @vyzo

File: timecache/first_seen_cache.go
Comment (smrz2001) [reply timecache/first_seen_cache.go:36]: What are your thoughts around where/how to do this? I could do it in this PR, if not complicated.

Or, if you would prefer to take some time and discuss the implementation, I could add a TODO for later.

If you meant in a goroutine with some kind of timed garbage collection, then would make sense to do in a separate PR.

File: timecache/last_seen_cache.go
Comment (smrz2001) [reply timecache/last_seen_cache.go:73]: Didn't do an `RLock` or even use a `RWMutex` (just using a plain `Mutex`) because the entry, if present and unexpired, would need to be "moved", which would need a write lock anyway.

Technically, I might have been able to do a `RLock`, checked for the entry and, if present, converted to `Lock` in order to move it.

From what I understand, "upgrading" locks like this without introducing race conditions is complicated and error-prone in Golang. Moreover, for an operation as small as delete + re-add (without GC), it felt like overkill.

This doesn't apply to the "first seen" cache implementation since it doesn't modify the cache when `Has` is called, unlike the "last seen" implementation.

File: timecache/first_seen_cache.go
Comment (vyzo) [reply timecache/first_seen_cache.go:36]: Lets add a TODO and do follow up, discussing in issue.

Comment (vyzo) [reply timecache/first_seen_cache.go:36]: and yes, a separate goroutine periodically running a sweep is what i have in mind.

File: timecache/last_seen_cache.go
Comment (vyzo) [reply timecache/last_seen_cache.go:73]: ok, good.

**Review by vyzo — APPROVED** [review]
ok, lets merge it and follow up for background gc.

**Comment by vyzo** [comment]
@lidel you are blocking merge.

**Comment by smrz2001** [comment]
> ok, lets merge it and follow up for background gc.

Great, thanks so much @vyzo and @lidel!

I opened #515 for the GC follow-up discussion.

**Pull request merged**
